app/src/org/commcare/tasks/DataPullTask.java

```diff
@@ -245,6 +225,57 @@ private static boolean shouldGenerateFirstKey() {
         return keyServer == null || keyServer.equals("");
     }
 
+    private ResultAndError<PullTaskResult> getRequestResultOrRetry(AndroidTransactionParserFactory factory) {
+        if (retryLimit != -1 && numTries >= retryLimit) {
+            // for testing purposes only
+            return new ResultAndError<>(PullTaskResult.RETRY_LIMIT_EXCEEDED);
+        }

```

Comment:
Instead of adding a custom `RETRY_LIMIT_EXCEEDED` return code for tests, can you try counting retry limit and number of tries in `HttpRequestEndpointsMock` and returning something like `DataPullTask.CONNECTION_TIMEOUT` in the mock and asserting it in the async test?

In the past I've added test code to the prod code base, but I think it is good to avoid whenever possible.